helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/AssignmentMetadataStore.java

```diff
@@ -61,36 +62,42 @@ protected AssignmentMetadataStore(BucketDataAccessor bucketDataAccessor, String
     _bestPossiblePath = String.format(BEST_POSSIBLE_TEMPLATE, clusterName, ASSIGNMENT_METADATA_KEY);
   }
 
-  public synchronized Map<String, ResourceAssignment> getBaseline() {
+  public Map<String, ResourceAssignment> getBaseline() {
     // Return the in-memory baseline. If null, read from ZK. This is to minimize reads from ZK
     if (_globalBaseline == null) {
-      try {
-        HelixProperty baseline =
-            _dataAccessor.compressedBucketRead(_baselinePath, HelixProperty.class);
-        _globalBaseline = splitAssignments(baseline);
-      } catch (ZkNoNodeException ex) {
-        // Metadata does not exist, so return an empty map
-        _globalBaseline = new HashMap<>();
+      // double-checked locking
+      synchronized (this) {
+        if (_globalBaseline == null) {
+          _globalBaseline = fetchAssignmentOrDefault(_baselinePath);
+        }
       }
     }
     return _globalBaseline;
   }
 
-  public synchronized Map<String, ResourceAssignment> getBestPossibleAssignment() {
+  public Map<String, ResourceAssignment> getBestPossibleAssignment() {
     // Return the in-memory baseline. If null, read from ZK. This is to minimize reads from ZK
     if (_bestPossibleAssignment == null) {
-      try {
-        HelixProperty baseline =
-            _dataAccessor.compressedBucketRead(_bestPossiblePath, HelixProperty.class);
-        _bestPossibleAssignment = splitAssignments(baseline);
-      } catch (ZkNoNodeException ex) {
-        // Metadata does not exist, so return an empty map
-        _bestPossibleAssignment = new HashMap<>();
+      // double-checked locking
+      synchronized (this) {
+        if (_bestPossibleAssignment == null) {
+          _bestPossibleAssignment = fetchAssignmentOrDefault(_bestPossiblePath);
+        }
       }
     }
     return _bestPossibleAssignment;
   }
 
+  private Map<String, ResourceAssignment> fetchAssignmentOrDefault(String path) {
+    try {
```

Comment:
Could you make it explicit in a comment that this function must be called in a synchronized setting? This is my worry in terms of "clarity".

Reply:
Thanks for the comment, but why this has to be synchronized? it's a read operation

Reply:
My reasoning was that this function was synchronized before your change (and even your change wraps this in a synchronized block), so it should continue to be synchronized down the road. 
You did raise a good point - my impression is bucket operations are not thread safe, but maybe we want to verify that.

Reply:
I took a look, I don't see a clear reason why this bucket read operation has to be synchronized. It might be true for set and delete. This is a private method after all, so the impact is limited. Let me know what you think on this. Thanks.